kyuubi-server/src/main/scala/org/apache/kyuubi/credentials/HadoopFsDelegationTokenProvider.scala

```diff
@@ -0,0 +1,109 @@
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.credentials.HadoopFsDelegationTokenProvider.{disableFsCache, doAsProxyUser}
+
+class HadoopFsDelegationTokenProvider extends HadoopDelegationTokenProvider with Logging {
+
+  override val serviceName: String = "hadoopfs"
+
+  override def delegationTokensRequired(
+      hadoopConf: Configuration,
+      kyuubiConf: KyuubiConf): Boolean = {
+    SecurityUtil.getAuthenticationMethod(hadoopConf) != AuthenticationMethod.SIMPLE
+  }
+
+  override def obtainDelegationTokens(
+      hadoopConf: Configuration,
+      kyuubiConf: KyuubiConf,
+      owner: String,
+      creds: Credentials): Unit = {
+    // FileSystem objects are cached in FileSystem.CACHE by a composite key.
+    // The UserGroupInformation object used to create it is part of that key.
+    // If cache is enabled, new FileSystem objects are created and cached at every method
+    // invocation.
+    val internalConf = disableFsCache(kyuubiConf, hadoopConf)
+
+    doAsProxyUser(owner) {
+      val fileSystems =
+        HadoopFsDelegationTokenProvider.hadoopFSsToAccess(kyuubiConf, internalConf)
+
+      try {
+        fileSystems.foreach { fs =>
+          info(s"getting token owned by $owner for: $fs")
+          fs.addDelegationTokens(null, creds)
+        }
+      } finally {
+        // Token renewal interval is longer than FileSystems' underlying connections' max idle time.
+        // Close FileSystems won't lose efficiency.
+        fileSystems.foreach(_.close())
+      }
+    }
+  }
+
+}
+
+object HadoopFsDelegationTokenProvider {
+
+  def disableFsCache(kyuubiConf: KyuubiConf, hadoopConf: Configuration): Configuration = {
+    // Avoid unnecessary disk io by not loading default resources
+    val newConf = new Configuration(false)
+    hadoopConf.iterator().asScala.foreach(e => newConf.set(e.getKey, e.getValue))
+
+    hadoopFSsToAccess(kyuubiConf, hadoopConf)
+      .foreach(fs => newConf.setBoolean(s"fs.${fs.getScheme}.impl.disable.cache", true))
+    newConf
+  }
+
+  def hadoopFSsToAccess(kyuubiConf: KyuubiConf, hadoopConf: Configuration): Set[FileSystem] = {
+    val defaultFS = FileSystem.get(hadoopConf)
+    val filesystemsToAccess = kyuubiConf
+      .get(KyuubiConf.CREDENTIALS_HADOOP_FS_URIS)
+      .map(new Path(_).getFileSystem(hadoopConf))
+      .toSet
+
+    filesystemsToAccess + defaultFS
+  }
+
+  def doAsProxyUser[T](proxyUser: String)(f: => T): T = {
+    // For some reason the Scala-generated anonymous class ends up causing an
+    // UndeclaredThrowableException, even if you annotate the method with @throws.
+    try {
+      val ugi = UserGroupInformation.createProxyUser(proxyUser, UserGroupInformation.getCurrentUser)
+      ugi.doAs(new PrivilegedExceptionAction[T] {
+        override def run(): T = f
+      })
+    } catch {
+      case e: UndeclaredThrowableException => throw Option(e.getCause).getOrElse(e)
```

Comment:
what happen if we fail to get token for one user, does it affects other user ?

Reply:
`HadoopCredentialsManager` will catch the exception and schedule a retry task. 

Failures are controlled in user scope and will not affect other users.